// @yusuke4869/mahjong/dist/mahjong/score/calculation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.calculateFinalScore = exports.calculateScore = void 0;
const convert_1 = require("./convert");
const calculateScore = ({ fu, han, tsumo = false, oya = false, yakuman = 0, kiriageMangan = false, sanma = false, }) => {
    let scoreType = null;
    let pay = { oya: null, ko: 0 };
    // 役満
    if (yakuman > 0) {
        scoreType = yakuman === 1 ? "役満" : `${(0, convert_1.num2Kanji)(yakuman)}倍役満`;
        pay = {
            oya: oya ? null : 16000 * yakuman,
            ko: 8000 * yakuman * (oya ? 2 : 1),
        };
    }
    // 数え役満
    else if (han >= 13) {
        scoreType = "数え役満";
        pay = {
            oya: oya ? null : 16000,
            ko: 8000 * (oya ? 2 : 1),
        };
    }
    // 三倍満
    else if (han >= 11 && han < 13) {
        scoreType = "三倍満";
        pay = {
            oya: oya ? null : 12000,
            ko: 6000 * (oya ? 2 : 1),
        };
    }
    // 倍満
    else if (han >= 8 && han < 11) {
        scoreType = "倍満";
        pay = {
            oya: oya ? null : 8000,
            ko: 4000 * (oya ? 2 : 1),
        };
    }
    // 跳満
    else if (han >= 6 && han < 8) {
        scoreType = "跳満";
        pay = {
            oya: oya ? null : 6000,
            ko: 3000 * (oya ? 2 : 1),
        };
    }
    // 満貫
    else if ((han >= 5 && han < 6) ||
        (fu >= 70 && han === 3) ||
        (fu >= 40 && han === 4) ||
        (kiriageMangan && fu >= 60 && han === 3) ||
        (kiriageMangan && fu >= 30 && han === 4)) {
        scoreType = "満貫";
        pay = {
            oya: oya ? null : 4000,
            ko: 2000 * (oya ? 2 : 1),
        };
    }
    // その他
    else {
        // スコアは 符 x 2^翻 x 4 で計算
        const score = fu * Math.pow(2, han) * 4;
        pay = {
            oya: oya ? null : score * 2,
            ko: score * (oya ? 2 : 1),
        };
    }
    const res = {
        scoreType: scoreType,
        score: (0, convert_1.pay2Score)(pay, tsumo, sanma),
        pay: tsumo
            ? {
                oya: pay.oya ? (0, convert_1.roundUp10)(pay.oya) : null,
                ko: (0, convert_1.roundUp10)(pay.ko),
            }
            : { oya: null, ko: null },
    };
    return res;
};
exports.calculateScore = calculateScore;
const calculateFinalScore = (rawScoreData, honba, sanma) => {
    const rawScore = rawScoreData.score;
    const res = {
        rawScore: rawScore,
        score: rawScore + 100 * (sanma ? 2 : 3) * honba,
        pay: {
            oya: rawScoreData.pay.oya ? rawScoreData.pay.oya + 100 * honba : null,
            ko: rawScoreData.pay.ko ? rawScoreData.pay.ko + 100 * honba : null,
        },
    };
    return res;
};
exports.calculateFinalScore = calculateFinalScore;
