// @yusuke4869/mahjong/dist/mahjong/score/score.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getScore = void 0;
const calculation_1 = require("./calculation");
const getScore = ({ fu, han, tsumo = false, oya = false, honba = 0, yakuman = 0, kazoeYakuman = true, kiriageMangan = false, sanma = false, }) => {
    const res = {
        fu: fu,
        han: han,
        scoreType: null,
        rawScore: 0,
        score: 0,
        pay: {
            oya: null,
            ko: null,
        },
        detail: {
            tsumo: tsumo,
            oya: oya,
            honba: honba,
        },
        rule: {
            kazoeYakuman: kazoeYakuman,
            kiriageMangan: kiriageMangan,
            sanma: sanma,
        },
    };
    // 数え役満がないとき
    if (!kazoeYakuman && han >= 13)
        han = 11;
    // 点数計算
    const rawScore = (0, calculation_1.calculateScore)({ han, fu, tsumo, oya, yakuman, kiriageMangan, sanma });
    res.scoreType = rawScore.scoreType;
    res.rawScore = rawScore.score;
    // 積み棒計算
    const result = (0, calculation_1.calculateFinalScore)(rawScore, honba, sanma);
    res.score = result.score;
    res.pay = result.pay;
    return res;
};
exports.getScore = getScore;
