// @yusuke4869/mahjong/dist/mahjong/agari/agari.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAgari = void 0;
const array_1 = require("../hai/array");
const convert_1 = require("./convert");
const mentsu_1 = require("./mentsu");
const toitsu_1 = require("./toitsu");
const validation_1 = require("./validation");
const getAgari = (hai, agariHai, tsumo) => {
    const validation = (0, validation_1.inputValidation)(hai);
    if (!validation)
        return null;
    const furo = (0, convert_1.furo2Elements)(hai.furo);
    if (!furo)
        return null;
    const aka = (0, convert_1.getAka)(hai.hai);
    const haiArrObj = (0, array_1.getHaiArrayObject)(hai.hai);
    const agariArr = [];
    const res = [];
    const toitsu = (0, toitsu_1.getToitsu)({
        elements: [],
        hai: {
            manzu: [...haiArrObj.manzu],
            pinzu: [...haiArrObj.pinzu],
            sozu: [...haiArrObj.sozu],
            jihai: [...haiArrObj.jihai],
        },
        completed: false,
    });
    if (toitsu.completed && toitsu.elements.length === 7) {
        const agari = {
            elements: toitsu.elements,
        };
        res.push((0, convert_1.replaceAka)(agari, aka));
    }
    for (const atama of toitsu.elements) {
        // 雀頭
        const r = (0, convert_1.initialize)(furo, haiArrObj, atama);
        if (!r)
            continue;
        // 刻子
        const comb = (0, mentsu_1.getKotsuCombination)(r, agariHai, tsumo !== null && tsumo !== void 0 ? tsumo : true);
        agariArr.push(...comb);
    }
    for (const v of agariArr) {
        // 順子
        const r = (0, mentsu_1.getShuntsu)(v);
        if (r.completed && r.elements.length === 5) {
            const agari = {
                elements: r.elements,
            };
            res.push((0, convert_1.replaceAka)(agari, aka));
        }
    }
    return res.length > 0 ? res : null;
};
exports.getAgari = getAgari;
