// @yusuke4869/mahjong/dist/mahjong/agari/convert.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.replaceAka = exports.getAka = exports.initialize = exports.furo2Elements = void 0;
const convert_1 = require("../hai/convert");
const validation_1 = require("./validation");
const furoType2ElementType = (furoType) => {
    switch (furoType) {
        case "chi":
            return "shuntsu";
        case "pon":
            return "minko";
        case "daiminkan":
            return "minkan";
        case "ankan":
            return "ankan";
        case "kakan":
            return "minkan";
    }
};
const furo2Elements = (furo) => {
    const elem = [];
    for (const f of furo) {
        const r = (0, validation_1.furoValidation)(f);
        if (!r)
            return false;
        elem.push({
            type: furoType2ElementType(f.type),
            hai: f.hai,
            furo: f.type === "ankan" ? false : true,
        });
    }
    return elem;
};
exports.furo2Elements = furo2Elements;
const initialize = (furo, haiArrObj, atama) => {
    if (atama.type !== "toitsu")
        return null;
    const r = {
        elements: [...furo],
        hai: {
            manzu: [...haiArrObj.manzu],
            pinzu: [...haiArrObj.pinzu],
            sozu: [...haiArrObj.sozu],
            jihai: [...haiArrObj.jihai],
        },
        completed: false,
    };
    r.elements.push(atama);
    const atamaHai = atama.hai[0];
    const n = (0, convert_1.getHaiNormalNum)(atamaHai);
    const t = (0, convert_1.getHaiSuffix)(atamaHai);
    r.hai[(0, convert_1.suffix2FullType)(t)][n - 1] -= 2;
    return r;
};
exports.initialize = initialize;
const getAka = (hai) => {
    const aka = [];
    for (const h of hai) {
        const num = parseInt(h);
        const t = (0, convert_1.getHaiSuffix)(h);
        if (num === 0 && (t === "m" || t === "p" || t === "s")) {
            aka.push(`5${t}`);
        }
    }
    return aka;
};
exports.getAka = getAka;
const replaceAka = (agari, aka) => {
    const res = {
        elements: [],
    };
    for (const e of agari.elements) {
        if (e.furo) {
            res.elements.push(e);
            continue;
        }
        const r = {
            type: e.type,
            hai: [],
            furo: e.furo,
        };
        for (const h of e.hai) {
            if (aka.includes(h)) {
                const t = (0, convert_1.getHaiSuffix)(h);
                r.hai.push((0, convert_1.getHai)(0, t));
                aka.splice(aka.indexOf(h), 1);
            }
            else
                r.hai.push(h);
        }
        res.elements.push(r);
    }
    return res;
};
exports.replaceAka = replaceAka;
