// @yusuke4869/mahjong/dist/mahjong/agari/mentsu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getShuntsu = exports.getKotsuCombination = void 0;
const array_1 = require("../hai/array");
const convert_1 = require("../hai/convert");
const getKotsuCombination = (data, agariHai, tsumo) => {
    const kotsu = [];
    const hai = {
        manzu: [...data.hai.manzu],
        pinzu: [...data.hai.pinzu],
        sozu: [...data.hai.sozu],
        jihai: [...data.hai.jihai],
    };
    for (const k of Object.keys(hai)) {
        const t = (0, convert_1.fullType2Suffix)(k);
        for (const [index, value] of data.hai[k].entries()) {
            if (value >= 3) {
                const n = index + 1;
                kotsu.push({
                    type: (0, convert_1.getHai)(n, t) === agariHai && !tsumo && value === 3 ? "minko" : "anko",
                    hai: [(0, convert_1.getHai)(n, t), (0, convert_1.getHai)(n, t), (0, convert_1.getHai)(n, t)],
                    furo: false,
                });
                hai[k][index] -= 3;
            }
        }
    }
    const res = [];
    // bit全探索
    for (let i = 0; i < 1 << kotsu.length; i++) {
        const r = {
            elements: [...data.elements],
            hai: {
                manzu: [...data.hai.manzu],
                pinzu: [...data.hai.pinzu],
                sozu: [...data.hai.sozu],
                jihai: [...data.hai.jihai],
            },
            completed: false,
        };
        for (let j = 0; j < kotsu.length; j++) {
            if ((i >> j) & 1) {
                r.elements.push(kotsu[j]);
                const hai = kotsu[j].hai[0];
                const t = (0, convert_1.getHaiSuffix)(hai);
                const n = (0, convert_1.getHaiNormalNum)(hai);
                r.hai[(0, convert_1.suffix2FullType)(t)][n - 1] -= 3;
            }
        }
        r.completed = (0, array_1.isHaiArrayObjectEmpty)(r.hai);
        res.push(r);
    }
    return res;
};
exports.getKotsuCombination = getKotsuCombination;
const getShuntsu = (data) => {
    if (data.completed)
        return data;
    const { elements, hai } = data;
    for (const k of Object.keys(hai)) {
        if (k === "jihai")
            continue;
        const t = (0, convert_1.fullType2Suffix)(k);
        for (let i = 0; i < 7;) {
            const v1 = hai[k][i];
            const v2 = hai[k][i + 1];
            const v3 = hai[k][i + 2];
            if (v1 > 0 && v2 > 0 && v3 > 0) {
                const n = i + 1;
                elements.push({
                    type: "shuntsu",
                    hai: [(0, convert_1.getHai)(n, t), (0, convert_1.getHai)(n + 1, t), (0, convert_1.getHai)(n + 2, t)],
                    furo: false,
                });
                hai[k][i]--;
                hai[k][i + 1]--;
                hai[k][i + 2]--;
                if (v1 > 0)
                    continue;
            }
            i++;
        }
    }
    const res = {
        elements: elements,
        hai: hai,
        completed: (0, array_1.isHaiArrayObjectEmpty)(hai),
    };
    return res;
};
exports.getShuntsu = getShuntsu;
