// @yusuke4869/mahjong/dist/mahjong/hai/convert.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fullType2Suffix = exports.suffix2FullType = exports.normal2Aka = exports.aka2Normal = exports.getHai = exports.getHaiSuffix = exports.getHaiNormalNum = void 0;
const getHaiNormalNum = (hai) => {
    const haiNum = parseInt(hai, 10);
    return haiNum === 0 ? 5 : haiNum;
};
exports.getHaiNormalNum = getHaiNormalNum;
const getHaiSuffix = (hai) => {
    return hai.slice(-1);
};
exports.getHaiSuffix = getHaiSuffix;
const getHai = (num, suffix) => {
    if ((num === 0 ||
        num === 1 ||
        num === 2 ||
        num === 3 ||
        num === 4 ||
        num === 5 ||
        num === 6 ||
        num === 7 ||
        num === 8 ||
        num === 9) &&
        (suffix === "m" || suffix === "p" || suffix === "s"))
        return `${num}${suffix}`;
    if ((num === 1 || num === 2 || num === 3 || num === 4 || num === 5 || num === 6 || num === 7) && suffix === "z")
        return `${num}${suffix}`;
    throw new Error(`Not Exists Hai: ${num}${suffix}`);
};
exports.getHai = getHai;
const aka2Normal = (hai) => {
    const num = parseInt(hai);
    const suffix = (0, exports.getHaiSuffix)(hai);
    if (num === 0 && (suffix === "m" || suffix === "p" || suffix === "s"))
        return (0, exports.getHai)(5, suffix);
    return hai;
};
exports.aka2Normal = aka2Normal;
const normal2Aka = (hai) => {
    const num = (0, exports.getHaiNormalNum)(hai);
    const suffix = (0, exports.getHaiSuffix)(hai);
    if (num === 5 && (suffix === "m" || suffix === "p" || suffix === "s"))
        return (0, exports.getHai)(0, suffix);
    return hai;
};
exports.normal2Aka = normal2Aka;
const suffix2FullType = (suffix) => {
    switch (suffix) {
        case "m":
            return "manzu";
        case "p":
            return "pinzu";
        case "s":
            return "sozu";
        case "z":
            return "jihai";
    }
};
exports.suffix2FullType = suffix2FullType;
const fullType2Suffix = (fullType) => {
    switch (fullType) {
        case "manzu":
            return "m";
        case "pinzu":
            return "p";
        case "sozu":
            return "s";
        case "jihai":
            return "z";
    }
};
exports.fullType2Suffix = fullType2Suffix;
