// @yusuke4869/mahjong/dist/mahjong/yaku/convert.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.doraHyo2Dora = exports.agariElementsRemoveAka = exports.notExistsRes = void 0;
const convert_1 = require("../hai/convert");
exports.notExistsRes = {
    exists: false,
    yaku: null,
    han: 0,
    yakuman: 0,
};
const agariElementsRemoveAka = (agari) => {
    const res = [];
    for (const e of agari) {
        const r = {
            type: e.type,
            hai: [],
            furo: e.furo,
        };
        for (const h of e.hai) {
            r.hai.push((0, convert_1.aka2Normal)(h));
        }
        res.push(r);
    }
    return res;
};
exports.agariElementsRemoveAka = agariElementsRemoveAka;
const doraHyo2Dora = (doraHyo) => {
    const haiNum = (0, convert_1.getHaiNormalNum)(doraHyo);
    const haiSuffix = (0, convert_1.getHaiSuffix)(doraHyo);
    if (haiSuffix === "z") {
        if (haiNum === 7)
            return (0, convert_1.getHai)(5, haiSuffix);
        else if (haiNum === 4)
            return (0, convert_1.getHai)(1, haiSuffix);
        return (0, convert_1.getHai)(haiNum + 1, haiSuffix);
    }
    if (haiNum === 9)
        return (0, convert_1.getHai)(1, haiSuffix);
    return (0, convert_1.getHai)(haiNum + 1, haiSuffix);
};
exports.doraHyo2Dora = doraHyo2Dora;
