// @yusuke4869/mahjong/dist/mahjong/yaku/ittsu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ittsu = void 0;
const convert_1 = require("../hai/convert");
const sort_1 = require("../hai/sort");
const convert_2 = require("./convert");
const ittsu = (data) => {
    let existsFuro = false;
    const haiArrObj = {
        manzu: [0, 0, 0, 0, 0, 0, 0, 0, 0],
        pinzu: [0, 0, 0, 0, 0, 0, 0, 0, 0],
        sozu: [0, 0, 0, 0, 0, 0, 0, 0, 0],
        jihai: [0, 0, 0, 0, 0, 0, 0],
    };
    for (const elem of data.agari.elements) {
        if (elem.furo)
            existsFuro = true;
        if (elem.type !== "shuntsu")
            continue;
        const t = (0, convert_1.suffix2FullType)((0, convert_1.getHaiSuffix)(elem.hai[0]));
        const hai = (0, sort_1.sortHai)(elem.hai);
        const n1 = (0, convert_1.getHaiNormalNum)(hai[0]);
        const n2 = (0, convert_1.getHaiNormalNum)(hai[1]);
        const n3 = (0, convert_1.getHaiNormalNum)(hai[2]);
        if (n1 === 1 || n1 === 4 || n1 === 7) {
            if (n2 === n1 + 1 && n3 === n1 + 2) {
                haiArrObj[t][n1 - 1]++;
                haiArrObj[t][n2 - 1]++;
                haiArrObj[t][n3 - 1]++;
            }
        }
    }
    for (const k of Object.keys(haiArrObj)) {
        if (haiArrObj[k].filter((v) => v !== 0).length === 9) {
            const res = {
                exists: true,
                yaku: "一気通貫",
                han: existsFuro ? 1 : 2,
                yakuman: 0,
            };
            return res;
        }
    }
    return convert_2.notExistsRes;
};
exports.ittsu = ittsu;
