// @yusuke4869/mahjong/dist/mahjong/yaku/pinfu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pinfu = exports.checkPinfu = void 0;
const convert_1 = require("./convert");
const waiting_1 = require("./waiting");
const checkPinfu = (data, bakaze, jikaze) => {
    const res = {
        exists: false,
        furo: false,
    };
    let furo = false;
    for (const elem of data.agari.elements) {
        if (elem.furo)
            furo = true;
        const t = elem.type;
        if (t === "toitsu") {
            const hai = elem.hai[0];
            if (hai === bakaze || hai === jikaze || hai === "5z" || hai === "6z" || hai === "7z")
                return res;
            continue;
        }
        if (t !== "shuntsu")
            return res;
    }
    const w = (0, waiting_1.waiting)(data);
    if (!w.ryammen)
        return res;
    res.exists = true;
    res.furo = furo;
    return res;
};
exports.checkPinfu = checkPinfu;
const pinfu = (data, bakaze, jikaze) => {
    const result = (0, exports.checkPinfu)(data, bakaze, jikaze);
    if (!result.exists)
        return convert_1.notExistsRes;
    if (result.furo)
        return convert_1.notExistsRes;
    const res = {
        exists: true,
        yaku: "平和",
        han: 1,
        yakuman: 0,
    };
    return res;
};
exports.pinfu = pinfu;
