// @yusuke4869/mahjong/dist/mahjong/yaku/waiting.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.waiting = void 0;
const convert_1 = require("../hai/convert");
const sort_1 = require("../hai/sort");
const convert_2 = require("./convert");
const waiting = (data) => {
    const tarElemArr = [];
    const agariHai = (0, convert_1.aka2Normal)(data.agariHai);
    for (const elem of (0, convert_2.agariElementsRemoveAka)(data.agari.elements)) {
        if (elem.furo)
            continue;
        if (elem.type === "ankan" || elem.type === "minkan")
            continue;
        // ロン和了の刻子は明刻になる
        if (elem.type === "anko" && !data.tsumo)
            continue;
        for (const hai of elem.hai) {
            if (hai === agariHai) {
                tarElemArr.push(elem);
                break;
            }
        }
    }
    const res = {
        ryammen: false,
        shampon: false,
        penchan: false,
        kanchan: false,
        tanki: false,
    };
    for (const elem of tarElemArr) {
        const t = elem.type;
        if (t === "toitsu") {
            res.tanki = true;
            continue;
        }
        if (t === "anko") {
            res.shampon = true;
            continue;
        }
        if (t !== "shuntsu")
            continue;
        const hai = (0, sort_1.sortHai)(elem.hai);
        if (hai[1] === agariHai) {
            res.kanchan = true;
            continue;
        }
        if (hai[0] === agariHai) {
            if ((0, convert_1.getHaiNormalNum)(agariHai) === 7)
                res.penchan = true;
            else
                res.ryammen = true;
            continue;
        }
        if (hai[2] === agariHai) {
            if ((0, convert_1.getHaiNormalNum)(agariHai) === 3)
                res.penchan = true;
            else
                res.ryammen = true;
        }
    }
    return res;
};
exports.waiting = waiting;
