// @yusuke4869/mahjong/dist/mahjong/yaku/yaku.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getYaku = void 0;
const agari_1 = require("../agari/agari");
const score_1 = require("../score/score");
const anko_1 = require("./anko");
const chanta_1 = require("./chanta");
const chitoi_1 = require("./chitoi");
const churempoto_1 = require("./churempoto");
const dora_1 = require("./dora");
const fu_1 = require("./fu");
const isshoku_1 = require("./isshoku");
const ittsu_1 = require("./ittsu");
const jihai_1 = require("./jihai");
const kantsu_1 = require("./kantsu");
const kokushi_1 = require("./kokushi");
const peko_1 = require("./peko");
const pinfu_1 = require("./pinfu");
const roto_1 = require("./roto");
const ryuiso_1 = require("./ryuiso");
const sangen_1 = require("./sangen");
const sanshokudojun_1 = require("./sanshokudojun");
const sanshokudoko_1 = require("./sanshokudoko");
const sushiho_1 = require("./sushiho");
const tanyao_1 = require("./tanyao");
const toitoi_1 = require("./toitoi");
const tsuiso_1 = require("./tsuiso");
const sortedYakuArr = [
    "天和",
    "地和",
    "大三元",
    "清老頭",
    "字一色",
    "緑一色",
    "小四喜",
    "大四喜",
    "四槓子",
    "四暗刻",
    "四暗刻単騎",
    "国士無双",
    "国士無双十三面待ち",
    "九蓮宝燈",
    "純正九蓮宝燈",
    "立直",
    "ダブル立直",
    "一発",
    "槍槓",
    "嶺上開花",
    "海底摸月",
    "河底撈魚",
    "門前清自摸和",
    "清一色",
    "純全帯么九",
    "二盃口",
    "混一色",
    "七対子",
    "対々和",
    "混老頭",
    "混全帯么九",
    "一気通貫",
    "小三元",
    "三槓子",
    "三暗刻",
    "三色同順",
    "三色同刻",
    "断么九",
    "平和",
    "一盃口",
    "白",
    "發",
    "中",
    "自風牌",
    "場風牌",
    "ドラ",
    "赤ドラ",
    "裏ドラ",
];
const getYaku = (data) => {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
    const doraHyo = (_a = data.doraHyo) !== null && _a !== void 0 ? _a : [];
    const uradoraHyo = (_b = data.uradoraHyo) !== null && _b !== void 0 ? _b : [];
    const tenho = (_c = data.tenho) !== null && _c !== void 0 ? _c : false;
    const oya = (_d = data.oya) !== null && _d !== void 0 ? _d : false;
    const honba = (_e = data.honba) !== null && _e !== void 0 ? _e : 0;
    const rempuhai = (_f = data.rempuhai) !== null && _f !== void 0 ? _f : 2;
    const kazoeYakuman = (_g = data.kazoeYakuman) !== null && _g !== void 0 ? _g : true;
    const doubleYakuman = (_h = data.doubleYakuman) !== null && _h !== void 0 ? _h : true;
    const kiriageMangan = (_j = data.kiriageMangan) !== null && _j !== void 0 ? _j : false;
    const sanma = (_k = data.sanma) !== null && _k !== void 0 ? _k : false;
    const res = {
        yaku: [],
        fu: 0,
        han: 0,
        yakuman: 0,
        scoreType: null,
        rawScore: 0,
        score: 0,
        pay: {
            oya: null,
            ko: null,
        },
        detail: {
            bakaze: data.bakaze,
            jikaze: data.jikaze,
            doraHyo: doraHyo,
            uradoraHyo: uradoraHyo,
            tsumo: data.tsumo,
            oya: oya,
            honba: honba,
            agari: [],
        },
        rule: {
            rempuhai: rempuhai,
            kazoeYakuman: kazoeYakuman,
            doubleYakuman: doubleYakuman,
            kiriageMangan: kiriageMangan,
            sanma: sanma,
        },
    };
    if (tenho) {
        res.yaku.push({ name: "天和", han: 0 });
        res.yakuman += 1;
    }
    else if (data.chiho) {
        res.yaku.push({ name: "地和", han: 0 });
        res.yakuman += 1;
    }
    const agariArr = (0, agari_1.getAgari)(data.hai, data.agariHai, data.tsumo);
    if (agariArr === null) {
        const kokushiRes = (0, kokushi_1.kokushi)({
            hai: data.hai,
            agari: {
                elements: [],
            },
            agariHai: data.agariHai,
            tsumo: data.tsumo,
            doubleYakuman: doubleYakuman,
            tenho: tenho,
        });
        if (kokushiRes.exists && kokushiRes.yaku) {
            res.yaku.push({ name: kokushiRes.yaku, han: kokushiRes.han });
            res.yakuman += kokushiRes.yakuman;
        }
        const score = (0, score_1.getScore)({
            fu: res.fu,
            han: res.han,
            tsumo: data.tsumo,
            oya: oya,
            honba: honba,
            yakuman: res.yakuman,
            kazoeYakuman: kazoeYakuman,
            kiriageMangan: kiriageMangan,
            sanma: sanma,
        });
        res.scoreType = score.scoreType;
        res.rawScore = score.rawScore;
        res.score = score.score;
        res.pay = score.pay;
        return res;
    }
    let max = {
        yaku: [],
        fu: 0,
        han: 0,
        yakuman: 0,
        agari: [],
    };
    for (const agari of agariArr) {
        const tmp = {
            yaku: [],
            fu: 0,
            han: 0,
            yakuman: 0,
            agari: agari.elements,
        };
        const input = {
            hai: data.hai,
            agari: agari,
            agariHai: data.agariHai,
            tsumo: data.tsumo,
            doubleYakuman: doubleYakuman,
            tenho: tenho,
        };
        const sangenRes = (0, sangen_1.sangen)(input);
        if (sangenRes.exists && sangenRes.yaku) {
            tmp.yaku.push({ name: sangenRes.yaku, han: sangenRes.han });
            tmp.han += sangenRes.han;
            tmp.yakuman += sangenRes.yakuman;
        }
        const rotoRes = (0, roto_1.roto)(input);
        if (rotoRes.exists && rotoRes.yaku) {
            tmp.yaku.push({ name: rotoRes.yaku, han: rotoRes.han });
            tmp.han += rotoRes.han;
            tmp.yakuman += rotoRes.yakuman;
        }
        const tsuisoRes = (0, tsuiso_1.tsuiso)(input);
        if (tsuisoRes.exists && tsuisoRes.yaku) {
            tmp.yaku.push({ name: tsuisoRes.yaku, han: tsuisoRes.han });
            tmp.han += tsuisoRes.han;
            tmp.yakuman += tsuisoRes.yakuman;
        }
        const ryuisoRes = (0, ryuiso_1.ryuiso)(input);
        if (ryuisoRes.exists && ryuisoRes.yaku) {
            tmp.yaku.push({ name: ryuisoRes.yaku, han: ryuisoRes.han });
            tmp.han += ryuisoRes.han;
            tmp.yakuman += ryuisoRes.yakuman;
        }
        const sushihoRes = (0, sushiho_1.sushiho)(input);
        if (sushihoRes.exists && sushihoRes.yaku) {
            tmp.yaku.push({ name: sushihoRes.yaku, han: sushihoRes.han });
            tmp.han += sushihoRes.han;
            tmp.yakuman += sushihoRes.yakuman;
        }
        const kantsuRes = (0, kantsu_1.kantsu)(input);
        if (kantsuRes.exists && kantsuRes.yaku) {
            tmp.yaku.push({ name: kantsuRes.yaku, han: kantsuRes.han });
            tmp.han += kantsuRes.han;
            tmp.yakuman += kantsuRes.yakuman;
        }
        const ankoRes = (0, anko_1.anko)(input);
        if (ankoRes.exists && ankoRes.yaku) {
            tmp.yaku.push({ name: ankoRes.yaku, han: ankoRes.han });
            tmp.han += ankoRes.han;
            tmp.yakuman += ankoRes.yakuman;
        }
        const churempotoRes = (0, churempoto_1.churempoto)(input);
        if (churempotoRes.exists && churempotoRes.yaku) {
            tmp.yaku.push({ name: churempotoRes.yaku, han: churempotoRes.han });
            tmp.han += churempotoRes.han;
            tmp.yakuman += churempotoRes.yakuman;
        }
        if (tmp.yakuman > 0 && tmp.yakuman > max.yakuman) {
            tmp.han = 0;
            max = tmp;
            continue;
        }
        const isshokuRes = (0, isshoku_1.isshoku)(input);
        if (isshokuRes.exists && isshokuRes.yaku) {
            tmp.yaku.push({ name: isshokuRes.yaku, han: isshokuRes.han });
            tmp.han += isshokuRes.han;
        }
        const chantaRes = (0, chanta_1.chanta)(input);
        // 混老頭とは複合しない
        if (!rotoRes.exists && chantaRes.exists && chantaRes.yaku) {
            tmp.yaku.push({ name: chantaRes.yaku, han: chantaRes.han });
            tmp.han += chantaRes.han;
        }
        const pekoRes = (0, peko_1.peko)(input);
        if (pekoRes.exists && pekoRes.yaku) {
            tmp.yaku.push({ name: pekoRes.yaku, han: pekoRes.han });
            tmp.han += pekoRes.han;
        }
        const chitoiRes = (0, chitoi_1.chitoi)(input);
        if (chitoiRes.exists && chitoiRes.yaku) {
            tmp.yaku.push({ name: chitoiRes.yaku, han: chitoiRes.han });
            tmp.han += chitoiRes.han;
        }
        const toitoiRes = (0, toitoi_1.toitoi)(input);
        if (toitoiRes.exists && toitoiRes.yaku) {
            tmp.yaku.push({ name: toitoiRes.yaku, han: toitoiRes.han });
            tmp.han += toitoiRes.han;
        }
        const ittsuRes = (0, ittsu_1.ittsu)(input);
        if (ittsuRes.exists && ittsuRes.yaku) {
            tmp.yaku.push({ name: ittsuRes.yaku, han: ittsuRes.han });
            tmp.han += ittsuRes.han;
        }
        const sanshokudojunRes = (0, sanshokudojun_1.sanshokudojun)(input);
        if (sanshokudojunRes.exists && sanshokudojunRes.yaku) {
            tmp.yaku.push({ name: sanshokudojunRes.yaku, han: sanshokudojunRes.han });
            tmp.han += sanshokudojunRes.han;
        }
        const sanshokudokoRes = (0, sanshokudoko_1.sanshokudoko)(input);
        if (sanshokudokoRes.exists && sanshokudokoRes.yaku) {
            tmp.yaku.push({ name: sanshokudokoRes.yaku, han: sanshokudokoRes.han });
            tmp.han += sanshokudokoRes.han;
        }
        const tanyaoRes = (0, tanyao_1.tanyao)(input);
        if (tanyaoRes.exists && tanyaoRes.yaku) {
            tmp.yaku.push({ name: tanyaoRes.yaku, han: tanyaoRes.han });
            tmp.han += tanyaoRes.han;
        }
        const pinfuRes = (0, pinfu_1.pinfu)(input, data.bakaze, data.jikaze);
        if (pinfuRes.exists && pinfuRes.yaku) {
            tmp.yaku.push({ name: pinfuRes.yaku, han: pinfuRes.han });
            tmp.han += pinfuRes.han;
        }
        const jihaiRes = (0, jihai_1.jihai)(input, data.bakaze, data.jikaze);
        for (const r of jihaiRes) {
            if (r.exists && r.yaku) {
                tmp.yaku.push({ name: r.yaku, han: r.han });
                tmp.han += r.han;
            }
        }
        tmp.fu = (0, fu_1.fu)(input, data.bakaze, data.jikaze, rempuhai);
        if (tmp.han >= max.han && tmp.fu >= max.fu) {
            max = tmp;
        }
    }
    if (max.yakuman === 0) {
        if (data.doubleReach) {
            max.yaku.push({ name: "ダブル立直", han: 2 });
            max.han += 2;
        }
        else if (data.reach) {
            max.yaku.push({ name: "立直", han: 1 });
            max.han++;
        }
        if (data.ippatsu) {
            max.yaku.push({ name: "一発", han: 1 });
            max.han++;
        }
        if (data.chankan && !data.tsumo) {
            max.yaku.push({ name: "槍槓", han: 1 });
            max.han++;
        }
        if (data.rinshan && data.tsumo) {
            max.yaku.push({ name: "嶺上開花", han: 1 });
            max.han++;
        }
        if (data.haitei && data.tsumo) {
            max.yaku.push({ name: "海底摸月", han: 1 });
            max.han++;
        }
        else if (data.hotei && !data.tsumo) {
            max.yaku.push({ name: "河底撈魚", han: 1 });
            max.han++;
        }
        if (data.tsumo && data.hai.furo.filter((v) => v.type !== "ankan").length === 0) {
            max.yaku.push({ name: "門前清自摸和", han: 1 });
            max.han++;
        }
        const doraRes = (0, dora_1.dora)({
            hai: data.hai,
            agari: {
                elements: [],
            },
            agariHai: data.agariHai,
            tsumo: data.tsumo,
            doubleYakuman: doubleYakuman,
            tenho: tenho,
        }, doraHyo, uradoraHyo);
        for (const r of doraRes) {
            if (r.exists && r.yaku) {
                max.yaku.push({ name: r.yaku, han: r.han });
                max.han += r.han;
            }
        }
    }
    res.yaku = max.yaku.sort((a, b) => sortedYakuArr.indexOf(a.name) - sortedYakuArr.indexOf(b.name));
    res.fu = max.fu;
    res.han = max.han;
    res.yakuman = max.yakuman;
    res.detail.agari = max.agari;
    const score = (0, score_1.getScore)({
        fu: res.fu,
        han: res.han,
        tsumo: data.tsumo,
        oya: oya,
        honba: honba,
        yakuman: res.yakuman,
        kazoeYakuman: kazoeYakuman,
        kiriageMangan: kiriageMangan,
        sanma: sanma,
    });
    res.scoreType = score.scoreType;
    res.rawScore = score.rawScore;
    res.score = score.score;
    res.pay = score.pay;
    return res;
};
exports.getYaku = getYaku;
